modules/ding_borgun_checkout/ding_borgun_checkout.module

```diff
@@ -0,0 +1,353 @@
+ * characters and can contain A-Z0-9.
+ *
+ * @return string|null
+ *   The order id, or null if locking failed.
+ */
+function _ding_borgun_checkout_get_order_id() {
+  $lock_id = 'ding_borgun_checkout_order_id';
+  // Use a lock to ensure that there's only one process that increment the order
+  // id at a time.
+  $locked = lock_acquire($lock_id);
+  if (!$locked) {
+    if (!lock_wait($lock_id, 30)) {
+      $locked = lock_acquire($lock_id);
+    }
+  }
+
+  if (!$locked) {
+    return NULL;
+  }
+
+  $counter = variable_get('ding_borgun_checkout_order_id_next', '1');
+  variable_set('ding_borgun_checkout_order_id_next', $counter + 1);
+  lock_release($lock_id);
+  $prefix = variable_get('ding_borgun_checkout_order_id_prefix', '');
+  // Using strlen is safe, we've limited the prefix to A-Z.
+  $padding = 12 - strlen($prefix);
+  return sprintf('%s%0' . $padding . 'd', $prefix, $counter);
+}
+
+/**
+ * Capture single use token.
+ *
+ * Authorizes and captures a payment using a single use token from
+ * Borgun.Checkout.
+ *
+ * @param string $token
+ *   Token supplied by Borgun.Checkout.
+ * @param int|float $amount
+ *   Payment amount.
+ * @param int $currency_code
+ *   The ISO 4217 currency code of the payment.
+ * @param string $order_id
+ *   Order ID of payment.
+ * @param string $meta
+ *   Additional information.
+ *
+ * @return string|false
+ *   Transaction id or false on failure.
+ */
+function _ding_borgun_checkout_capture_token($token, $amount, $currency_code, $order_id, $meta) {
+  $body = array(
+    "TransactionType" => "Sale",
+    // Amount with 2 decimal points, as an integer. JPY is an exception, but
+    // we're not supporting that.
+    "Amount" => floor($amount * 100),
+    // ISO 4217 currency code.
+    "Currency" => $currency_code,
+    // ISO 8601.
+    "TransactionDate" => date("c", REQUEST_TIME),
+    "PaymentMethod" => array(
+      "PaymentType" => "TokenSingle",
+      "Token" => $token,
+    ),
+    "OrderId" => $order_id,
+    "Metadata" => array(
+      "Payload" => $meta,
+    ),
+
+  );
+
+  $private_token = variable_get('ding_borgun_checkout_private_token', '');
+
+  $options = array(
+    'auth' => array($private_token, ''),
+    'base_uri' => _ding_borgun_checkout_host() . '/rpg/api/',
+  );
+  $client = new HttpClient($options);
+
+  try {
+    $response = $client->post('payment', array('json' => $body));
+  }
+  catch (ClientException $e) {
+    switch ($e->getCode()) {
+      case 400:
+        $error = json_decode((string) $e->getResponse()->getBody());
+        watchdog('ding_borgun_checkout', 'Form error in capture: @error',
+          array('@error' => $error->Message), WATCHDOG_ERROR);
+        break;
+
+      case 401:
+        watchdog('ding_borgun_checkout', 'Error in capture: Unauthorized',
+          array(), WATCHDOG_ERROR);
+        break;
+
+      default:
+    }
+
+    return FALSE;
+  }
+  catch (Exception $e) {
+    if ($e instanceof RequestException && $e->getResponse()) {
```

Comment:
Given the `if` statement why is this not `catch (RequestException $e)`? Is there a situation where catching the generic Exception makes sense?

Reply:
Missing a few `return FALSE;` there. Yes, we wan't to catch all exceptions (something went wrong). We could split it up in two catches, however, RequestException might not have a response, in which case we'll want to log it like a regular Exception. Hence the if.